source.go

```diff
@@ -83,161 +231,53 @@ func generateSourceFromImage(s *Spec, name string, st llb.State, cmd *Command, s
 
 		rOpts = append(rOpts, withConstraints(opts))
 		cmdSt := st.Run(rOpts...)
-		out = cmdSt.AddMount(subPath, out)
+
+		// on first iteration with a root subpath
+		// do not use AddMount, as this will overwrite / with a
+		// scratch fs
+		if i == 0 && subPath == "/" {
+			out = cmdSt.Root()
+		} else {
+			out = cmdSt.AddMount(subPath, out)
+		}
 	}
 
 	return out, nil
 }
 
-func Source2LLBGetter(s *Spec, src Source, name string) LLBGetter {
-	return source2LLBGetter(s, src, name, false)
-}
-
-// isRootPath is used to encapsulate various different possibilities for what amounts to the root path.
-// It helps prevent making an extra copy of the source when it is not necessary.
-func isRootPath(p string) bool {
-	return p == "" || p == "/" || p == "."
-}
-
-func needsFilter(o *filterOpts) bool {
-	if !isRootPath(o.source.Path) && !o.forMount && !o.pathHandled {
-		return true
-	}
-	if o.includeExcludeHandled {
-		return false
+func Source2LLBGetter(_ *Spec, src Source, name string, sOpt SourceOpts, opts ...llb.ConstraintsOpt) (llb.State, error) {
```

Comment:
`Source2LLBGetter` is the only caller of `GetSource` and the only caller of `GetFilter`.
I don't think we'd ever want to call `GetSource` without `GetFilter`.

Probably that should be in `GetSource`.

Reply:
I guess for code readability its nice to have some things split, but we also don't need to export those functions.

Maybe a `GetSource` (or maybe even `func(s *Source) AsState(opts ...llb.ConstraintsOpt) (llb.state, error)`) that calls `getSource().With(getFilter())`

Reply:
Yeah, I also like to have things split because of the mounting case. I.e., when we are mounting that should only affect the filter, so I wasn't a fan of passing the `forMount` flag into `getSource`. But I've wrapped `getSource` and `filter` into a single function.